Add proposal for secret and configmap files mode bits

rata opened the pull request:
This is a proposal to address https://github.com/kubernetes/kubernetes/issues/28317.

cc @pmorie (owner) @thockin 

@thockin: Sorry if you preferred not to be CCed, I thougth you'd be interested :-)

I think this is always the case, but let me say it one more time just in case: as this is a PR, ALL the feedback is more than welcome!

It's my first time in kubernetes, so sorry in advance if this is obviously wrong. What I realize now is that I forgot to add the headers to the proposal. Is there some script to add them? Or should I just c&p from some other proposal?

Thanks a lot,
Rodrigo

rata committed:
Add proposal for secret and configmap files mode bits

This is a proposal to address
https://github.com/kubernetes/kubernetes/issues/28317.

rata commented:
Ohhh, probably the filed should be  permission` and inside there can be  `mode` now and in the future `uid` and `gid` too.
Does it seem better with that change?

zhouhaibing089 commented:
Run `./hack/update-generated-docs.sh` would give you the headers.

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,111 @@
+
+The proposition is to add it as a field in the KeyToPath. That is, as
+v1.SecretVolumeSource in the API[5] and the equivalent for configmaps[4]. And
+then change the key to be optional, but is mandatory to specify at least one of
```
I don't understand why the key would become optional?  In general you can not take a field that used to be required and make it optional, anyway, since a client is allowed to assume it was always populated (we said it was required) and fail if it is not.   Now you're changing that promise.

It's unfortunate that this is tied to KeyToPath projections, which are not always used.  It might make sense to add a `defaultMode` to `type SecretVolumeSource` and `type ConfigMapVolumeSource`, too.  That way, really simple cases can just set the default, but complex cases can override that.

rata replied:
I was thinking about the Path, sorry. But yes, you are right, it may break assumptions doing that.

What I had in mind was:
If you create a secret/configmap from a dir, for example, you may just want to change the permissions on a key file. But having to specify the path, if it is the same as the key and you don't need anything but the identity mapping, is disturbing.

Your suggestion seems better. Although, I'm not 100% happy with having to specify the same path as the key, if you just want to change the permissions of one file.

Do you have any other great idea? :)

thockin replied:
I don't have a better answer that allows identity-mapping keys-to-files, allows single-file overrides AND stays compatible.  We are not going to embed file info into the Secret itself.  It seems like a corner-case that isn't so bad

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,111 @@
+   mappings and doesn't seem like a big issue.
+
+This setting might conflict with the `fsGroup` specified, and in that case if
+the `mode` bits are present for a file the `fsGroup` will be ignored. The reason
```
I don't think you need or want to ignore fsGroup.  It's possible to set the permissions such that you can't read them, but we can't really defend against that..

thockin replied:
I do think you should spec that we AND the provided mask with 0777 before applying it.  I don't think we should allow users to set sticky and setuid bits, without fully understanding the use case.  We can revisit that later if anyone complains.

rata replied:
Didn't follow the last part of the first comment. I don't need/want to ignore `fsGroup`, but do you think it should be done, then?

thockin replied:
Your doc said "if the `mode` bits are present for a file the `fsGroup` will be ignored".  I am saying I don't see why fsGroup would be ignored.  Yes, it is possible to set the mode on a file such that you can't read it (e.g. 0000) but we can't stop that.

rata replied:
Ohh, not sure what you want the end result to be. Do you want that the file mode is set on creation of the file (AtomicWriter now) and don't alter the rest, so `volume.SetVolumeOwnership` is called at the end as it is now?

thockin replied:
yes.  They are overlapping features, but that's just what it is.  It is less surprising that the features both work as described (and maybe they intersect badly) than to have one disable the other.

rata replied:
On Sun, Jul 10, 2016 at 09:55:19PM -0700, Tim Hockin wrote:
> > +Adding it there allows the user to change the mode bits of every file in the
> > +object, so it achieves the goal.
> > +
> > +The are two downside:
> > +
> > + * The files are symlinks pointint to the real file, and the realfile
> > +   permissions are only set. The symlink has the clasic symlink permissions.
> > +   This is something already present in 1.3, and it seems applications like ssh
> > +   work just fine with that. Something worth mentioning, but doesn't seem to be
> > +   an issue.
> > + * If the secret/configMap is mounted in more than one container, the file
> > +   permissions will be the same on all. This is already the case for Key
> > +   mappings and doesn't seem like a big issue.
> > +
> > +This setting might conflict with the `fsGroup` specified, and in that case if
> > +the `mode` bits are present for a file the `fsGroup` will be ignored. The reason
> 
> yes.  They are overlapping features, but that's just what it is.  It is less surprising that the features both work as described (and maybe they intersect badly) than to have one disable the other.

Ohh, you are right. This seems better :-)

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,111 @@
+
+# Proposed implementation
+
+The proposed implementation goes with the first alternative: adding a `mode` to
```
I assume you mean an integer, but you should specify.

rata replied:
Yes, I mean an integer. I will add it, thanks! :-)

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,111 @@
+
+The proposition is to add it as a field in the KeyToPath. That is, as
+v1.SecretVolumeSource in the API[5] and the equivalent for configmaps[4]. And
+then change the key to be optional, but is mandatory to specify at least one of
```
You can not change key to be optional.  That is a breaking change.  We just can't do it.

thockin commented:
Let me know when the default mode fields are specced and this is ready for another round.

rata commented:
@thockin thanks a lot for the feedback! Added it to the proposal :)

rata commented:
@zhouhaibing089 thanks! Sadly my notebook started swaping and it was unusable. Will try again in a while closing X, maybe that helps :)

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,151 @@
+        // the '..' path or start with '..'.
+        Items       []KeyToPath `json:"items,omitempty"`
+        // Mode bits to use on created files by default
+        DefaultMode int         `json:"defaultMode,omitempty"`
```
TBD: "mode" vs "defaultMode".

TBD: does this apply to intermediate directories created by file projections?  Directories need X bits set to be listable.  I would rather avoid adding a second mode field for directories, but we need to spec the behavior.

rata replied:
On Sun, Jul 10, 2016 at 10:04:02PM -0700, Tim Hockin wrote:
> > +In other words, the fields will look like this (comments might be improved in
> > +the implementation):
> > +```
> > +type SecretVolumeSource struct {
> > +        // Name of the secret in the pod's namespace to use.
> > +        SecretName string `json:"secretName,omitempty"`
> > +        // If unspecified, each key-value pair in the Data field of the referenced
> > +        // Secret will be projected into the volume as a file whose name is the
> > +        // key and content is the value. If specified, the listed keys will be
> > +        // projected into the specified paths, and unlisted keys will not be
> > +        // present. If a key is specified which is not present in the Secret,
> > +        // the volume setup will error. Paths must be relative and may not contain
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> 
> TBD: "mode" vs "defaultMode".

I like a little more `defaultMode` as it is clear that if you specify one in an
item (items[0].mode) that will be used for that file. If both are named `mode`
it seems kind of contradiction and not sure which will win (the volume-level
mode or the more specific one)

What do you think?

> TBD: does this apply to intermediate directories created by file projections?  Directories need X bits set to be listable.  I would rather avoid adding a second mode field for directories, but we need to spec the behavior.

Ohh, yes, I was thinking in files only and have a quick&dirty first
implementation working for files only. I will make it more clear when I come
back home :-)

rata replied:
Ohh, now that I read your comment again I'm not sure if you prefer this field to affect the directory or not.
My idea was that this setting didn't affect directories and only affects files. `fsGroup` can still be used and you usually don't want the same permissions on dirs and files for the reason you said.

So, the `defaultMode`, IMHO, shouldn't affect directories.

What do you think?

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,151 @@
+        // May not start with the string '..'.
+        Path string `json:"path"`
+        // Mode bits to use on this file
+        Mode int `json:"mode,omitempty"`
```
Same comment on directories - we need to spec what directory modes will be.

rata replied:
On Sun, Jul 10, 2016 at 10:04:33PM -0700, Tim Hockin wrote:
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> > +}
> > +
> > +type KeyToPath struct {
> > +        // The key to project.
> > +        Key string `json:"key"`
> > +
> > +        // The relative path of the file to map the key to.
> > +        // May not be an absolute path.
> > +        // May not contain the path element '..'.
> > +        // May not start with the string '..'.
> > +        Path string `json:"path"`
> > +        // Mode bits to use on this file
> > +        Mode int `json:"mode,omitempty"`
> 
> Same comment on directories - we need to spec what directory modes will be.

Yes, will add it. Thanks a lot, nice catch! :-)

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,151 @@
+        // May not contain the path element '..'.
+        // May not start with the string '..'.
+        Path string `json:"path"`
+        // Mode bits to use on this file
```
comment on the 0777 mask

rata replied:
On Sun, Jul 10, 2016 at 10:04:49PM -0700, Tim Hockin wrote:
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> > +}
> > +
> > +type KeyToPath struct {
> > +        // The key to project.
> > +        Key string `json:"key"`
> > +
> > +        // The relative path of the file to map the key to.
> > +        // May not be an absolute path.
> > +        // May not contain the path element '..'.
> > +        // May not start with the string '..'.
> > +        Path string `json:"path"`
> > +        // Mode bits to use on this file
> 
> comment on the 0777 mask

I realize about the same when going to sleep. Will add it, thanks a lot!

rata commented:
@thockin PTAL. I addressed the comments and specify that it doesn't apply for directories. But I'm not sure we want that. And if we want that, we might want to rename `defaultMode`  to something more specific to files.
But I'm not sure if having it only for files will be enough and need for directory perms will come later. I'll play with `fsGroup` and see if there is room to set dir perms with that usefully.

@nowprovision what was your use case? If you have some time, can you explain it in detail and comment if you need to set permission on directories and how?

Thanks!

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,159 @@
+        // Mode bits to use on created files by default. The used mode bits will
+        // be the provided AND 0777.
+        // Directories within the path are not affected by this setting
+        DefaultMode int         `json:"defaultMode,omitempty"`
```
I still want to think about naming here, as you suggest, but we can proceed to impl at this point

rata replied:
On Mon, Jul 11, 2016 at 10:27:42PM -0700, Tim Hockin wrote:
> > +```
> > +type SecretVolumeSource struct {
> > +        // Name of the secret in the pod's namespace to use.
> > +        SecretName string `json:"secretName,omitempty"`
> > +        // If unspecified, each key-value pair in the Data field of the referenced
> > +        // Secret will be projected into the volume as a file whose name is the
> > +        // key and content is the value. If specified, the listed keys will be
> > +        // projected into the specified paths, and unlisted keys will not be
> > +        // present. If a key is specified which is not present in the Secret,
> > +        // the volume setup will error. Paths must be relative and may not contain
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default. The used mode bits will
> > +        // be the provided AND 0777.
> > +        // Directories within the path are not affected by this setting
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> 
> I still want to think about naming here, as you suggest, but we can proceed to impl at this point

Awesome! I have some WIP that is quite close already :)

thockin commented on docs/proposals/secret-configmap-file-mode.md:
```diff
@@ -0,0 +1,159 @@
+        // be the provided AND 0777.
+        // Directories within the path are not affected by this setting
+        DefaultMode int         `json:"defaultMode,omitempty"`
+}
```
We should do the same in DownwardAPIVolumeSource

thockin replied:
and DownwardAPIVolumeFile

rata replied:
On Mon, Jul 11, 2016 at 10:28:34PM -0700, Tim Hockin wrote:
> > +
> > +type ConfigMapVolumeSource struct {
> > +        LocalObjectReference `json:",inline"`
> > +        // If unspecified, each key-value pair in the Data field of the referenced
> > +        // ConfigMap will be projected into the volume as a file whose name is the
> > +        // key and content is the value. If specified, the listed keys will be
> > +        // projected into the specified paths, and unlisted keys will not be
> > +        // present. If a key is specified which is not present in the ConfigMap,
> > +        // the volume setup will error. Paths must be relative and may not contain
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default. The used mode bits will
> > +        // be the provided AND 0777.
> > +        // Directories within the path are not affected by this setting
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> > +}
> 
> We should do the same in DownwardAPIVolumeSource

Ohh, didn't know if we wanted it for DownwardAPIVolumeSource. I realized the 3
share a lot while starting on the impl yesterday. I'll add it, then :)

rata replied:
On Mon, Jul 11, 2016 at 10:28:57PM -0700, Tim Hockin wrote:
> > +
> > +type ConfigMapVolumeSource struct {
> > +        LocalObjectReference `json:",inline"`
> > +        // If unspecified, each key-value pair in the Data field of the referenced
> > +        // ConfigMap will be projected into the volume as a file whose name is the
> > +        // key and content is the value. If specified, the listed keys will be
> > +        // projected into the specified paths, and unlisted keys will not be
> > +        // present. If a key is specified which is not present in the ConfigMap,
> > +        // the volume setup will error. Paths must be relative and may not contain
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default. The used mode bits will
> > +        // be the provided AND 0777.
> > +        // Directories within the path are not affected by this setting
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> > +}
> 
> and DownwardAPIVolumeFile

Sure! I'll check if it is needed in some other place too, just in case :-)

thockin commented:
LGTM.  Can you also cover downward API?

rata commented:
On Mon, Jul 11, 2016 at 10:29:25PM -0700, Tim Hockin wrote:
> LGTM.  Can you also cover downward API?

Thanks! Yes, will add it! :-)

rata commented:
@thockin one option we may want to consider for directory permissions is using the existing dir and it's permissions if it already exists. I'll like to play with it a little more before changing anything, but seems like a reasonable trade-off and easy to implement, it seems. Right now it is using 0777 for dirs (os.ModePerm) and that can still be the case if the dir doesn't exist, maybe.

rata commented:
@thockin added downwardAPI now :) Added it as another commit so it's easier for you to see the diff. Also, added you as an author, only me was not the right thing :-)

rata commented:
@thockin oh, and one more question. There are a few cleanups that I think are nice in the code I'm changing now. Should I do them after the implementation for this is merged, in a different PR? Should I do them in the same PR, before the actual implementation? Is there any convention on this?
I tend to do them before the actual implementation (to avoid coding stuff just to change them later), but don't know if there is any pref on this project :)

rata commented:
@thockin oh, and revisiting all the comments I'm not sure if maybe a "FileInfo" type, that ¿later? can have uid and guid, isn't maybe better? something like @pmorie says here: https://github.com/kubernetes/kubernetes/issues/4789#issuecomment-160051733

Or something like I said here: https://github.com/kubernetes/kubernetes/pull/28733#issuecomment-231557505

I have an implementation working (although write permission on group and others is removed somewhere down the line, I need to check why is that happening)  but it's of course trivial to adapt. And always better to have a nice API :-)

rata commented:
Sorry to bother again, but just seen this email that there would be a (not yet defined, but close) deadline for proposals to make it into 1.4: https://groups.google.com/forum/#!topic/kubernetes-pm/Ki63EztfZMo

I, of course, want to keep the API as clean as possible. So whatever you think is the best, it's fine for me too. If you are too busy with other stuff to make proper review before the deadline, I understand :-)

thockin commented:
@k8s-bot test this: github issue #IGNORE

thockin commented:
run hack/update-munge-docs.sh

rata commented:
@thockin run that, but  k8s-merge-robot removed LGTM. Can you please LGTM again? :-) Sorry, didn't know that will happen

thockin commented:
ok to test

pmorie commented:
Note to self: review this

pmorie commented:
Just a note: @rata I'm sorry I didn't review this while I was open -- I was sick recently and missed quite a bit of work around that and a vacation I couldn't reschedule.

@thockin @smarterclayton shall we do this for 1.4?

rata commented:
On Thu, Jul 28, 2016 at 01:08:19PM -0700, Paul Morie wrote:
> Just a note: @rata I'm sorry I didn't review this while I was open -- I was sick recently and missed quite a bit of work around that and a vacation I couldn't reschedule.

Ohh, no problem. Hope you are fine now.

> 
> @thockin @smarterclayton shall we do this for 1.4?

I think it was the plan, yes. If you are ok with the proposal, please review the
PR that implements this :-)
